flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/arrow/stream/StreamArrowPythonGroupWindowAggregateFunctionOperator.java

```diff
@@ -0,0 +1,547 @@
+	 * returned.
+	 *
+	 * @param window the window whose cleanup time we are computing.
+	 */
+	private long cleanupTime(W window) {
+		if (windowAssigner.isEventTime()) {
+			long cleanupTime = window.maxTimestamp() + allowedLateness;
+			return cleanupTime >= window.maxTimestamp() ? cleanupTime : Long.MAX_VALUE;
+		} else {
+			return window.maxTimestamp();
+		}
+	}
+
+	private void emitWindowResult(W window) throws Exception {
+		windowAccumulateData.setCurrentNamespace(window);
+		windowRetractData.setCurrentNamespace(window);
+		Iterable<RowData> currentWindowAccumulateDatas = windowAccumulateData.get();
+		if (currentWindowAccumulateDatas != null) {
+			currentBatchCount = 0;
+			for (RowData accumulateData : currentWindowAccumulateDatas) {
+				if (!hasRetractData(accumulateData)) {
+					arrowSerializer.write(getFunctionInput(accumulateData));
+					currentBatchCount++;
+				}
+			}
+			if (currentBatchCount > 0) {
+				inputKeyAndWindow.add(Tuple2.of((RowData) getCurrentKey(), window));
+				arrowSerializer.finishCurrentBatch();
+				pythonFunctionRunner.process(baos.toByteArray());
+				elementCount += currentBatchCount;
+				checkInvokeFinishBundleByCount();
+				currentBatchCount = 0;
+				baos.reset();
+			}
+		}
+	}
+
+	private boolean hasRetractData(RowData accumulateData) throws Exception {
+		BinaryRowData binaryAccumulateRowData = (BinaryRowData) accumulateData;
+		Iterable<RowData> currentWindowRetractDatas = windowRetractData.get();
+		if (currentWindowRetractDatas != null) {
+			for (RowData retractData : currentWindowRetractDatas) {
+				if (retractData.getRowKind() == RowKind.UPDATE_BEFORE) {
+					retractData.setRowKind(RowKind.UPDATE_AFTER);
+				} else {
+					retractData.setRowKind(RowKind.INSERT);
+				}
+				BinaryRowData binaryRetractData = (BinaryRowData) retractData;
+				if (binaryAccumulateRowData.getSizeInBytes() == binaryRetractData.getSizeInBytes() &&
+					BinaryRowDataUtil.byteArrayEquals(
+						binaryAccumulateRowData.getSegments()[0].getHeapMemory(),
+						binaryRetractData.getSegments()[0].getHeapMemory(),
+						binaryAccumulateRowData.getSizeInBytes())) {
+					return true;
+				}
+
+			}
+		}
+		return false;
+	}
+
+	/**
+	 * Registers a timer to cleanup the content of the window.
+	 *
+	 * @param window the window whose state to discard
+	 */
+	private void registerCleanupTimer(W window) {
+		long cleanupTime = cleanupTime(window);
+		if (cleanupTime == Long.MAX_VALUE) {
+			// don't set a GC timer for "end of time"
+			return;
+		}
+
+		if (windowAssigner.isEventTime()) {
+			triggerContext.registerEventTimeTimer(cleanupTime);
+		} else {
+			triggerContext.registerProcessingTimeTimer(cleanupTime);
+		}
+	}
+
+	private void setWindowProperty(W currentWindow) {
+		for (int i = 0; i < namedProperties.length; i++) {
+			switch (namedProperties[i]) {
+				case 0:
+					windowProperty.setField(i, TimestampData.fromEpochMillis(((TimeWindow) currentWindow).getStart()));
```

Comment:
Why it's always TimeWindow?

Reply:
Because CountWindow won't set the start property or the end property, we can make sure it is a TimeWindow.